packages/hardhat-core/test/internal/util/download.ts

```diff
@@ -0,0 +1,117 @@
+import { assert } from "chai";
+import fsExtra from "fs-extra";
+import path from "path";
+// @ts-ignore
+// tslint:disable-next-line no-implicit-dependencies
+import Proxy from "proxy";
+
+import { download } from "../../../src/internal/util/download";
+import { useTmpDir } from "../../helpers/fs";
+
+describe("Compiler List download", function () {
+  useTmpDir("compiler-downloader");
+
+  describe("basic", function () {
+    it("Should call download with the right params", async function () {
+      const compilersDir = this.tmpDir;
+      const downloadPath = path.join(compilersDir, "downloadedCompiler");
+      const expectedUrl = `https://solc-bin.ethereum.org/wasm/list.json`;
+
+      // download the file
+      await download(expectedUrl, downloadPath);
+      // Assert that the file exists
+      assert.isTrue(await fsExtra.pathExists(downloadPath));
+    });
+  });
+
+  describe("with proxy", function () {
+    let oldHttpsProxyEnv: string | undefined;
+    let oldHttpProxyEnv: string | undefined;
+    let proxy: any;
+    let proxyPort: number;
+    let connections = 0;
+
+    beforeEach(function (done) {
+      // Setup Proxy Server
+      proxy = new Proxy();
+      connections = 0;
+      proxy.on("connection", () => {
+        connections += 1;
+      });
+      proxy.listen(function () {
+        proxyPort = proxy.address().port;
+        done();
+      });
+    });
+
+    afterEach(function (done) {
+      proxy.close(done);
+    });
+
+    describe("Compilers list download with HTTPS_PROXY", function () {
+      beforeEach(function () {
+        oldHttpsProxyEnv = process.env.HTTPS_PROXY;
+        process.env.HTTPS_PROXY = `http://127.0.0.1:${proxyPort}`;
+      });
+
+      it("Should call download with the right params", async function () {
+        const compilersDir = this.tmpDir;
+        const downloadPath = path.join(compilersDir, "downloadedCompilerProxy");
+        const expectedUrl = `https://solc-bin.ethereum.org/wasm/list.json`;
+
+        // download the file
+        assert.equal(connections, 0);
+        await download(expectedUrl, downloadPath);
+        assert.equal(connections, 1);
+
+        // Assert that the file exists
+        assert.isTrue(await fsExtra.pathExists(downloadPath));
+      });
+
+      afterEach(function () {
+        if (oldHttpsProxyEnv !== undefined) {
+          process.env.HTTPS_PROXY = oldHttpsProxyEnv;
+        } else {
+          delete process.env.HTTPS_PROXY;
+        }
```

Comment:
To save the current environment, you were doing something like this:

```
env = process.env
process.env.HTTP_PROXY = ...
```

But since `process.env` is an object, any modifications you were doing to it were also done to `env`, so nothing was restored afterwards.

I also have to explicitly use `delete` here because it turns out that

```
process.env.FOO = undefined
```

doesn't remove the `FOO` environment variable. Instead, it sets it to the string `"undefined"`. Terrible behavior from node, but it is what it is.

Reply:
Nice catch. I think adapted that portion from the `https-proxy-agent` package, and in fact I wasn't sure if the `env` really needed to be changed back at the end of the test because I wasn't sure if changes to the `process.env` were scoped to the test or not. It seems like they are. 

Thanks for catching this.